sygnal/firebasepushkin.py

```diff
@@ -13,259 +13,419 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import json
 import logging
-from typing import Dict, Optional
+from functools import partial
+from typing import Dict
 
 import attr
-from firebase_admin import credentials, initialize_app, messaging
+from firebase_admin import (
+    credentials,
+    exceptions as firebase_exceptions,
+    initialize_app,
+    messaging,
+)
+from opentracing import logs
 from prometheus_client import Histogram
-from twisted.internet.defer import Deferred
-from twisted.python.threadpool import ThreadPool
 
-from .exceptions import PushkinSetupException
+from sygnal.utils import NotificationLoggerAdapter, twisted_sleep
+
+from .exceptions import (
+    NotificationDispatchException,
+    PushkinSetupException,
+    TemporaryNotificationDispatchException,
+)
 from .notifications import Pushkin
 
 SEND_TIME_HISTOGRAM = Histogram(
     "sygnal_fcm_request_time", "Time taken to send HTTP request"
 )
 
-MAX_TRIES = 3
-RETRY_DELAY_BASE = 10
-MAX_BYTES_PER_FIELD = 1024
-DEFAULT_MAX_CONNECTIONS = 20
+NOTIFICATION_DATA_INCLUDED = [
+    "type",
+    "room_id",
+    "event_id",
+    "sender_display_name",
+]
+
+DEFAULT_HANDLER = {"m.room.message": "message", "m.call.invite": "voip"}
 
 logger = logging.getLogger(__name__)
 
 
 @attr.s
 class FirebaseConfig(object):
     credentials = attr.ib()
-    max_connections = attr.ib(default=20)
     message_types = attr.ib(default=attr.Factory(dict), type=Dict[str, str])
+    event_handlers = attr.ib(default=attr.Factory(dict), type=Dict[str, str])
+    android_click_action = attr.ib(default=None, type=str)
 
 
 class FirebasePushkin(Pushkin):
+
+    MAX_TRIES = 3
+    RETRY_DELAY_BASE = 10
+
     def __init__(self, name, sygnal, config):
         super(FirebasePushkin, self).__init__(name, sygnal, config)
 
-        self.db = sygnal.database
-        self.reactor = sygnal.reactor
         self.config = FirebaseConfig(
             **{x: y for x, y in self.cfg.items() if x != "type"}
         )
 
+        self._app = initialize_app(self._load_credentials(), name="app")
+
+    def _load_credentials(self):
         credential_path = self.config.credentials
         if not credential_path:
             raise PushkinSetupException("No Credential path set in config")
 
-        cred = credentials.Certificate(credential_path)
+        return credentials.Certificate(credential_path)
 
-        self._pool = ThreadPool(maxthreads=self.config.max_connections)
-        self._pool.start()
+    async def _dispatch_message(self, n, data, device, span):
+        """
+        Construct Firebase message and dispatch to device.
 
-        self._app = initialize_app(cred, name="app")
+        Args:
+            n: The notification for the user and device.
```

Comment:
What types are these arguments (see other comment for formatting example)?

Missing `device` and `span`.